aggregator_core/src/datastore.rs

```diff
@@ -3658,10 +3660,40 @@ impl<C: Clock> Transaction<'_, C> {
                     &i64::try_from(batch.outstanding_aggregation_jobs())?,
                     /* client_timestamp_interval */
                     &SqlInterval::from(batch.client_timestamp_interval()),
+                    /* now */ &self.clock.now().as_naive_date_time()?,
                 ],
             )
-            .await?,
-        )
+            .await?;
+        let row_count = rows.len().try_into()?;
+
+        if let Some(row) = rows.into_iter().next() {
+            // We got a row back, meaning we wrote a batch. We check that it wasn't expired per the
+            // task's report_expiry_age, but otherwise we are done.
+            if row.get("is_expired") {
```

Comment:
Since this overall insertion function is relatively simple ([as compared to this](https://github.com/divviup/janus/pull/1526/files#diff-ad4962043a4d91715b26c52cd1ac0b490a8ac647f82f7cf34dd61f56ff134972R1565)), can't we check whether the batch is expired outside of SQL and just no-op?

Reply:
We could do that, but to do so, we'd need to know the task's report expiry age (since expiry is determined by this parameter).

Since this method doesn't take a `Task`, we'd need to read the report expiry age, inducing an extra round-trip for each insertion. With the strategy I take here, a non-expired insertion is one round-trip, and an expired insert is two round-trips. In the (hopeful) case that the large majority of inserts are non-expired, this is a savings.

There are a few alternate strategies I could have taken:
* I could take the `Task` (or just the expiry) as an argument. I _think_ this would work out in the sense that I think application uses of these methods either have or could plumb the relevant `Task` data through to the insertion method calls; but this would complicate the usage of every insertion method, and we'd need to deal with (unlikely, natch) issues of passing the wrong task.
* I could skip expiry-checking on insert at all, always writing a row to the DB. We would then rely on the expiry-checking on read to avoid reading expired data, and lean on the GC to lazily delete these rows eventually. (TBH, maybe I should have gone this way...)

Reply:
Hmm taking another look with a fresh brain I'm not sure why I thought this was doable without an extra round-trip, it's right there `(SELECT report_expiry_age FROM tasks WHERE task_id = $1)`.

> I could skip expiry-checking on insert at all, always writing a row to the DB. We would then rely on the expiry-checking on read to avoid reading expired data, and lean on the GC to lazily delete these rows eventually. (TBH, maybe I should have gone this way...)

I like that but I don't find enough rationale to suggest executing on it.